<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/Assignment%202a/Rojak/Group_Rojak_File_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Assignment: Alternatives to Pandas for Processing Large Datasets (Vaex)
###Group Rojak Members: 

<table>
  <tr>
    <th>Name</th>
    <th>Matric No</th>
  </tr>
  <tr>
    <th>Eddie Wong Chung Pheng </th>
    <th>A20EC0031</th>
  </tr>
  <tr>
    <th>Madihah Binti Che Zabri </th>
    <th>A20EC0074</th>
  </tr>
  <tr>
    <th>Nurarissa Dayana Binti Mohd Sukri</th>
    <th>A20EC0120</th>
  </tr>
  <tr> 
    <th>Vincent Boo Ee Khai</th>
    <th>A20EC0231</th>
  </tr>
</table>
<br></br>

####File 1: Explanation and Implementation of the basic concept of Vaex

For processing large datasets, Vaex Python's Out of Core Dataframe is a faster alternative to Pandas. It also has quick, user-friendly visualization tools. Pandas are typically used by Python developers when working with dataframes and processing. Its widespread acceptance can be attributed to its wide range of useful features and the simplicity of its application programming interface (API). However, pandas do not come without flaws, and Vaex can be used as an alternative. 

In this study, we will use the Vaex basic library to manage a large dataset and conduct basic data analysis in order to provide insights on the data. Rate.csv from the Health Insurance Marketplace will be used as the dataset.
<br></br>
**Dataset**
The dataset is obtained from Kaggle: <a href="https://www.kaggle.com/datasets/hhs/health-insurance-marketplace?select=Rate.csv">Rate.csv</a>  <br>
The size of the dataset is <b> 1.97 GB</b>.

#### Attribute Information:
| Acronym | Description |
| --- | --- |
| **BusinessYear** |   The year for which the rate information applies.  |
|**StateCode** |  The two-letter code for the state in which the health insurance plan is offered.  |
| **IssuerId** | A unique identifier for the insurer offering the health insurance plan. |
| **SourceName** |  The source of the rate information (e.g. the insurer, the state insurance department). |
| **VersionNum** | A version number for the rate information.  |
| **ImportDate** |  The date on which the rate information was imported into the Marketplace database.   |
| **IssuerId2** | A unique identifier for the insurer offering the health insurance plan.  |
| **FederalTIN** | Federal income taxes  |
| **RateEffectiveDate** |  The date for which the rate information is effective.   |
|**RateExpirationDate** |  The expire date for the rate. |
| **PlanId** | A unique identifier for the health insurance plan. |
| **RatingAreaId** | The age of the insured person for which the rate information applies.  |
| **Tobacco** | The rate information applies to tobacco users or non-tobacco users. |
| **Age** |   The age of the insured person for which the rate information applies.  |
| **IndividualRate** |  The monthly premium (cost) for the health insurance plan for an individual.  |
| **IndividualTobaccoRate** | The monthly premium for the health insurance plan for an individual tobacco user.  |
| **Couple** | The monthly premium for the health insurance plan for a couple.  |
|**PrimarySubscriberAndOneDependent** |  The primary subscriber for the health insurance plan and one dependent. |
| **PrimarySubscriberAndTwoDependents** | The primary subscriber for the health insurance plan and two dependent. |
| **CoupleAndOneDependent** | The monthly premium for the health insurance plan for a couple and one dependent. |
| **CoupleAndTwoDependents** | The monthly premium for the health insurance plan for a couple and two dependents.  |
| **CoupleAndThreeOrMoreDependents** |  The monthly premium for the health insurance plan for a couple and three or more dependents.   |
| **RowNumber** | The row number of rate information.  |



---



## **Installation and Downloading the Dataset**

**1) Install Vaex and Ipython**

Using Google Colab to run Vaex increases the likelihood of encountering an out-of-date IPython version error; to prevent this, we will update the IPython.

In [ ]:
!pip install vaex
!pip install ipython -U

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


**2) Import Vaex Library**

Vaex is a Python library for exploring large tabular datasets with the help of lazy Out-of-Core DataFrames. In a single second, it can process up to a billion objects/rows in an N-dimensional grid and calculate statistical measures like mean, sum, count, and standard deviation. Histograms, density plots, and 3D volume rendering are used for visualization, making big data accessible for interactive exploration. Vaex achieves its high performance through the use of memory mapping, a zero memory copy policy, and lazy computations so that no memory is wasted.

In [ ]:
import vaex as vx

**3) Download Dataset Using Google Drive**

This 1.97 GB Rate.csv dataset originates from the Health Insurance Marketplace.

In [ ]:
# Read dataset from google drive
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
link = 'https://drive.google.com/file/d/10PKFCHEW_B_4alRrDpjbA1WOsG2h4anZ/view?usp=share_link'
# to get the id part of the file
id = link.split("/")[-2]
 
downloaded = drive.CreateFile({'id':id})
downloaded.GetContentFile('Rate.csv') 

## **Data Cleaning and Preparation using Vaex**

**1) Read data with Vaex**

Vaex is compatible with HDF5 and CSV files. The vaex.from csv function is available for importing CSV data. In addition to reading CSV files, this function can also convert them to HDF5 format. It can be converted to HDF5 format by entering convert=True. Vaex will read the CSV in chunks, then convert each chunk to a temporary HDF5 file, and finally combine all of these files into a single HDF5 file. Using the chunk size argument, you can specify the size of each chunk. Here, we use the %%time magic command to measure how long an operation takes to complete. Processing time for Vaex.from_csv() is 1 minutes 54 seconds.

In [ ]:
%time df = vx.from_csv('Rate.csv', convert=True)

CPU times: user 31 ms, sys: 0 ns, total: 31 ms
Wall time: 32.4 ms


You can open a dataframe with the Open function by specifying its path. Using the Open function, you can convert your dataframe to other formats, such as HDF5. This function can read the dataset from a path or a list of paths.

In [ ]:
%time df2 = vx.open('Rate.csv')

CPU times: user 1.4 s, sys: 893 ms, total: 2.29 s
Wall time: 6.79 s


Hierarchical Data Format 5 (HDF5) is a free and open file format that can handle large, complex, and varied datasets. HDF5 employs a "file directory" structure similar to that of a computer's file system, allowing for flexible data organization within the file. We can see that converting the file to HDF5 saves a significant amount of time (in milliseconds) when compared to not converting it.

In [ ]:
%%time

df3 = vx.open('Rate.csv.hdf5')

CPU times: user 25.3 ms, sys: 2.77 ms, total: 28.1 ms
Wall time: 33.1 ms


**2) View Dataset**

Before analyzing the data, we can look at the top 5 records in the data set and read the summary and description. 

In [ ]:
df.head(5)

#    BusinessYear  StateCode      IssuerId  SourceName      VersionNum  ImportDate             IssuerId2  FederalTIN    RateEffectiveDate    RateExpirationDate    PlanId          RatingAreaId    Tobacco        Age              IndividualRate    IndividualTobaccoRate    Couple    PrimarySubscriberAndOneDependent    PrimarySubscriberAndTwoDependents    PrimarySubscriberAndThreeOrMoreDependents    CoupleAndOneDependent    CoupleAndTwoDependents    CoupleAndThreeOrMoreDependents    RowNumber
  0            2014  AK                21989  HIOS                     6  2014-03-19 07:06:49        21989  93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  0-20                      29                         nan     nan                                nan                                  nan                                          nan                      nan                       nan                               nan              14
  1            2014  AK                21989  HIOS                     6  2014-03-19 07:06:49        21989  93-0438772    2014-01-01           2014-12-31            21989AK0020001  Rating Area 1   No Preference  Family Option             36.95                      nan      73.9                              107.61                               107.61                                       107.61                   144.56                    144.56                            144.56           14
  2            2014  AK                21989  HIOS                     6  2014-03-19 07:06:49        21989  93-0438772    2014-01-01           2014-12-31            21989AK0020001  Rating Area 2   No Preference  Family Option             36.95                      nan      73.9                              107.61                               107.61                                       107.61                   144.56                    144.56                            144.56           15
  3            2014  AK                21989  HIOS                     6  2014-03-19 07:06:49        21989  93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  21                        32                         nan     nan                                nan                                  nan                                          nan                      nan                       nan                               nan              15
  4            2014  AK                21989  HIOS                     6  2014-03-19 07:06:49        21989  93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  22                        32                         nan     nan                                nan                                  nan                                          nan                      nan                       nan                               nan              16

In [ ]:
df.info()

The dataset has 12694445 records and 24 attributes.

In [ ]:
df.shape

(12694445, 24)

In [ ]:
df.describe()

BusinessYear StateCode           IssuerId SourceName  \
data_type               int64    string              int64     string   
count                12694445  12694445           12694445   12694445   
NA                          0         0                  0          0   
mean       2015.0335246637408        --  52485.92425151316         --   
std                  0.794052        --       26412.625647         --   
min                      2014        --              10046         --   
max                      2016        --              99969         --   

                  VersionNum ImportDate          IssuerId2 FederalTIN  \
data_type              int64     string              int64     string   
count               12694445   12694445           12694445   12694445   
NA                         0          0                  0          0   
mean       6.865558281594824         --  52485.92425151316         --   
std                  3.85718         --       26412.625647         --   
min                        1         --              10046         --   
max                       24         --              99969         --   

          RateEffectiveDate RateExpirationDate  ...     IndividualRate  \
data_type            string             string  ...            float64   
count              12694445           12694445  ...           12694445   
NA                        0                  0  ...                  0   
mean                     --                 --  ...  4098.026458607817   
std                      --                 --  ...       61222.710175   
min                      --                 --  ...                0.0   
max                      --                 --  ...           999999.0   

          IndividualTobaccoRate            Couple  \
data_type               float64           float64   
count                   4932349             40941   
NA                      7762096          12653504   
mean          543.6910791972604  48.1157143206081   
std                  294.591552         30.422519   
min                       41.73               0.0   
max                     6604.61             182.4   

          PrimarySubscriberAndOneDependent PrimarySubscriberAndTwoDependents  \
data_type                          float64                           float64   
count                                40941                             40941   
NA                                12653504                          12653504   
mean                     49.83590093060736                 70.66001196844134   
std                              29.691623                         44.619585   
min                                    0.0                               0.0   
max                                 169.65                            251.74   

          PrimarySubscriberAndThreeOrMoreDependents CoupleAndOneDependent  \
data_type                                   float64               float64   
count                                         40941                 40941   
NA                                         12653504              12653504   
mean                              89.40549400356075      72.7827932879011   
std                                       57.698574             48.231948   
min                                             0.0                   0.0   
max                                          357.94                260.85   

          CoupleAndTwoDependents CoupleAndThreeOrMoreDependents  \
data_type                float64                        float64   
count                      40941                          40941   
NA                      12653504                       12653504   
mean           89.01285874795374             107.80777362545656   
std                    59.138371                      73.240279   
min                          0.0                            0.0   
max                       339.31                         449.14   

                   RowNumber  
dat

**3) Filter Dataset**

In Python, filtering is one of the most time-consuming steps. But filtering the Vaex dataframe takes very little time. This is due to Vaex's zero memory copy policy. This means that filtering a DataFrame requires minimal memory and does not duplicate the data. df_filtered has a "view" on the original data. Even when you filter a 1TB file, only a fraction will be read. This means that when you have a large number of missing values, you can either drop them or fill them at almost no cost.

In [ ]:
%time df_filtered = df[df['BusinessYear']==2016]  # easily filter your DataFrame, without making a copy
df_filtered.head(5)

CPU times: user 4.15 ms, sys: 14.9 ms, total: 19 ms
Wall time: 27.5 ms


#    BusinessYear  StateCode      IssuerId  SourceName      VersionNum  ImportDate             IssuerId2  FederalTIN    RateEffectiveDate    RateExpirationDate    PlanId          RatingAreaId    Tobacco        Age              IndividualRate    IndividualTobaccoRate    Couple    PrimarySubscriberAndOneDependent    PrimarySubscriberAndTwoDependents    PrimarySubscriberAndThreeOrMoreDependents    CoupleAndOneDependent    CoupleAndTwoDependents    CoupleAndThreeOrMoreDependents    RowNumber
  0            2016  AK                21989  HIOS                     4  2015-08-22 15:09:32        21989  93-0438772    2016-01-01           2016-12-31            21989AK0030001  Rating Area 1   No Preference  0-20                      43                         nan    nan                                 nan                                  nan                                          nan                      nan                       nan                               nan              14
  1            2016  AK                21989  HIOS                     4  2015-08-22 15:09:32        21989  93-0438772    2016-01-01           2016-03-31            21989AK0080001  Rating Area 1   No Preference  Family Option             50.67                      nan    100.33                              116.04                               116.04                                       116.04                   117.86                    117.86                            117.86           14
  2            2016  AK                21989  HIOS                     4  2015-08-22 15:09:32        21989  93-0438772    2016-04-01           2016-06-30            21989AK0080001  Rating Area 1   No Preference  Family Option             51.36                      nan    101.7                               117.62                               117.62                                       117.62                   180.28                    180.28                            180.28           14
  3            2016  AK                21989  HIOS                     4  2015-08-22 15:09:32        21989  93-0438772    2016-01-01           2016-03-31            21989AK0090001  Rating Area 1   No Preference  0-20                      61                         nan    nan                                 nan                                  nan                                          nan                      nan                       nan                               nan              14
  4            2016  AK                21989  HIOS                     4  2015-08-22 15:09:32        21989  93-0438772    2016-04-01           2016-06-30            21989AK0090001  Rating Area 1   No Preference  0-20                      62                         nan    nan                                 nan                                  nan                                          nan                      nan                       nan                               nan              14

**4) Find and Handle Missing Values**

The ismissing(), isna(), and isnan() functions can be used to obtain information regarding missing data. The first function returns true when missing values (masked arrays), missing strings, or None are encountered. The second function returns a boolean expression that indicates whether the specified values are Not Available (missing or NaN). The third function returns an array containing values of type NaN.

In [ ]:
%time df.IndividualTobaccoRate.ismissing().sum()

CPU times: user 94.6 ms, sys: 2.81 ms, total: 97.4 ms
Wall time: 66.4 ms


array(0)

In [ ]:
%time df.IndividualTobaccoRate.isna().sum()

CPU times: user 126 ms, sys: 1.71 ms, total: 128 ms
Wall time: 84.2 ms


array(7762096)

In [ ]:
%time df.IndividualTobaccoRate.isnan().sum()

CPU times: user 125 ms, sys: 0 ns, total: 125 ms
Wall time: 78 ms


array(7762096)

In the IndividualTobaccoRate column of the results, there are 7762096 values of "Not Available" or NaN. Then, we use fillna with the value 0 to replace any NaN values. Instead of omitting the column, which would remove a huge record of data, we can assume that NaN indicates that the person in question did not smoke.

In [ ]:
%time df_fillna = df.fillna(value=0, column_names=['IndividualTobaccoRate'])

df_fillna

CPU times: user 2.41 ms, sys: 826 µs, total: 3.24 ms
Wall time: 3.26 ms


#           BusinessYear    StateCode    IssuerId    SourceName    VersionNum    ImportDate           IssuerId2    FederalTIN    RateEffectiveDate    RateExpirationDate    PlanId          RatingAreaId    Tobacco        Age            IndividualRate    IndividualTobaccoRate    Couple    PrimarySubscriberAndOneDependent    PrimarySubscriberAndTwoDependents    PrimarySubscriberAndThreeOrMoreDependents    CoupleAndOneDependent    CoupleAndTwoDependents    CoupleAndThreeOrMoreDependents    RowNumber
0           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  0-20           29.0              0.0                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan                               14
1           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0020001  Rating Area 1   No Preference  Family Option  36.95             0.0                      73.9      107.61                              107.61                               107.61                                       144.56                   144.56                    144.56                            14
2           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0020001  Rating Area 2   No Preference  Family Option  36.95             0.0                      73.9      107.61                              107.61                               107.61                                       144.56                   144.56                    144.56                            15
3           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  21             32.0              0.0                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan                               15
4           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  22             32.0              0.0                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan                               16
...         ...             ...          ...         ...           ...           ...                  ...          ...           ...                  ...                   ...             ...             ...            ...            ...               ...                      ...       ...                                 ...                                  ...                                          ...                      ...                       ...                               ...
12,694,440  2016            WV           96480       SERFF         2             2015-08-20 12:28:36  96480        13-5123390    2016-01-01           2016-12-31            96480WV0090003  Rating Area 11  No Preference  61             14.05             0.0                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan                               2033
12,694,441  2016            WV           96480

**5) Drop Unnecessary Column**

Dropna(), dropnan(), dropmissing(), and dropinf() can be used to remove null values (). The functions all accomplish the same things, but they all drop by in different ways. The dropna() and dropmissing() functions, for instance, both remove values by employing the isna() and ismissing() filters, respectively. We did not eliminate any columns due to their lack of data because some records do not use the corresponding column. Instead, we choose to eliminate the unnecessary column.

In [ ]:
%time df.drop('RowNumber')

CPU times: user 2.61 ms, sys: 912 µs, total: 3.52 ms
Wall time: 4.23 ms


#           BusinessYear    StateCode    IssuerId    SourceName    VersionNum    ImportDate           IssuerId2    FederalTIN    RateEffectiveDate    RateExpirationDate    PlanId          RatingAreaId    Tobacco        Age            IndividualRate    IndividualTobaccoRate    Couple    PrimarySubscriberAndOneDependent    PrimarySubscriberAndTwoDependents    PrimarySubscriberAndThreeOrMoreDependents    CoupleAndOneDependent    CoupleAndTwoDependents    CoupleAndThreeOrMoreDependents
0           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  0-20           29.0              nan                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan
1           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0020001  Rating Area 1   No Preference  Family Option  36.95             nan                      73.9      107.61                              107.61                               107.61                                       144.56                   144.56                    144.56
2           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0020001  Rating Area 2   No Preference  Family Option  36.95             nan                      73.9      107.61                              107.61                               107.61                                       144.56                   144.56                    144.56
3           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  21             32.0              nan                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan
4           2014            AK           21989       HIOS          6             2014-03-19 07:06:49  21989        93-0438772    2014-01-01           2014-12-31            21989AK0010001  Rating Area 1   No Preference  22             32.0              nan                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan
...         ...             ...          ...         ...           ...           ...                  ...          ...           ...                  ...                   ...             ...             ...            ...            ...               ...                      ...       ...                                 ...                                  ...                                          ...                      ...                       ...
12,694,440  2016            WV           96480       SERFF         2             2015-08-20 12:28:36  96480        13-5123390    2016-01-01           2016-12-31            96480WV0090003  Rating Area 11  No Preference  61             14.05             nan                      nan       nan                                 nan                                  nan                                          nan                      nan                       nan
12,694,441  2016            WV           96480       SERFF         2             2015-08-20 12:28:36  96480        13-5123390    2016-01-01           2016-12-31            96480WV0090003  Rating Area 11  No Preference  62             14.05             nan                      nan       

**6) Count Values**

We can also count non-null values of a choosen attributes using count while value_counts will give both NaN and non-nulls.

In [ ]:
df.count('Couple') #counts not-null values

array(40941)

In [ ]:
df['PrimarySubscriberAndOneDependent'].value_counts()

NaN      12653504
0.00         1848
7.35         1037
4.41          929
7.72          834
           ...   
43.39           1
58.03           1
94.61           1
65.44           1
39.59           1
Length: 5975, dtype: int64

**7) Selection**

A Selection defines a subset of data. First, it expedites the process of filtering data from the dataframe. In addition, selections permit the calculation of statistics for multiple subsets in a single pass through the data. Multiple actions can be performed on a single line. This application is especially helpful when working with DataFrames that do not fit in memory. Due to the fact that Vaex does not duplicate data but instead keeps track of the rows that have been selected internally, the execution time is extremely low (in the millisecond range).

In [ ]:
df.mean(df.IndividualRate, selection=[df.Tobacco=='No Preference'])

array([6381.26425215])

**8) Groupby / Aggregations**

Vaex implements parallelized, highly performant groupby operations, especially when using categories (>1 billion/second).



In [ ]:
df.groupby(by='StateCode').agg({'IndividualRate':'mean'})

#    StateCode    IndividualRate
0    NJ           422.017152475854
1    OK           5157.578136289506
2    SC           5681.339818376415
3    NE           9155.720852636192
4    IN           4855.9184819586535
...  ...          ...
34   NC           8206.358099093733
35   AK           7024.934674098514
36   ME           355.5591839399675
37   IL           372.44531510277506
38   HI           256.34147562582365

##Conclusion

In this study, we learned the fundamentals of Vaex, a Python library for the rapid processing of large datasets that can be a suitable alternative to Pandas, particularly for large datasets. Vaex possesses a large number of additional capabilities and functions. Vaex is not only a viable alternative to Pandas, but it also provides features such as memory efficiency, easy parallelization, visualization tools, and a sophisticated query language. Memory mapping, a zero memory copy policy, and lazy computations are utilized by Vaex to achieve optimal performance.

##Reference

1. https://www.kaggle.com/datasets/hhs/health-insurance-marketplace?datasetId=21&sortBy=voteCount&select=Rate.csv
2. https://vaex.io/docs/index.html
3. https://www.machinelearningplus.com/python/vaex/